amaranth/vendor/_lattice_nexus.py

```diff
@@ -0,0 +1,884 @@
+
+    Build products:
+        * ``{{name}}_impl/{{name}}_impl.htm``: consolidated log.
+        * ``{{name}}.bit``: binary bitstream.
+        * ``{{name}}.xcf``: JTAG programming vector. (if using ``programmer``)
+    """
+
+    toolchain = None # selected when creating platform
+
+    device  = abstractproperty()
+    package = abstractproperty()
+    speed   = abstractproperty()
+    grade   = "C" # [C]ommercial, [I]ndustrial
+
+    # Oxide templates
+
+    _nextpnr_device_options = { # TODO: Add Certus-NX devices
+        "LIFCL-40":    "--40k",
+        "LIFCL-17":    "--17k",
+    }
+
+    _nextpnr_package_options = { # TODO: Add Certus-NX packages
+        "SG72": "QFN72",
+        "UWG72": "WLCSP72",
+        "MG121": "csfBGA121",
+        "BG196": "caBGA196",
+        "BG256": "caBGA256",
+        "MG289": "csBGA289",
+        "BG400": "caBGA400",
+    }
+
+    _oxide_required_tools = [ 
+        "yosys",
+        "nextpnr-nexus",
+        "prjoxide"
+    ]
+    _oxide_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "{{name}}.il": r"""
+            # {{autogenerated}}
+            {{emit_rtlil()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ys": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v") -%}
+                read_verilog {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".sv") -%}
+                read_verilog -sv {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".il") -%}
+                read_ilang {{file}}
+            {% endfor %}
+            read_ilang {{name}}.il
+            delete w:$verilog_initial_trigger
+            {{get_override("script_after_read")|default("# (script_after_read placeholder)")}}
+            synth_nexus {{get_override("synth_opts")|options}} -top {{name}}
+            {{get_override("script_after_synth")|default("# (script_after_synth placeholder)")}}
+            write_json {{name}}.json
+        """,
+        "{{name}}.pdc": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                ldc_set_location -site {{ '{' }}{{pin_name}}{{ '}' }} {{'['}}get_ports {{port_name}}{{']'}} 
+                {% if attrs -%}
+                ldc_set_port -iobuf {{ '{' }}{%- for key, value in attrs.items() %}{{key}}={{value}} {% endfor %}{{ '}' }} {{'['}}get_ports {{port_name}}{{']'}}
+                {% endif %}
+            {% endfor %}
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
+            {#
+                {% if port_signal is not none -%}
+                    set_frequency "{{port_signal.name}}" {{frequency/1000000}};
```

Comment:
Why does this have different syntax than the below? Is one of the commands broken?

Reply:
`set_frequency` is not supported by nextpnr-nexus, so this has been removed in https://github.com/amaranth-lang/amaranth/pull/759/commits/8ccc083c024f48991b2218ba57ba6bb91e33bc49 (tbh I don't remember why the syntax differs, or where `set_frequency` came from).

Reply:
I don't think that's legitimate--how would nextpnr know the timing target for your design? Or, equally, how do you know that your design will function with the clock you give it?

Reply:
I tried looking into how to pass timing info to nextpnr and according to [nextpnr/docs/nexus.md] (https://github.com/YosysHQ/nextpnr/blob/master/docs/nexus.md), "Timing constraints are currently ignored, but should be expected to be supported soon.". A grep of `set_frequency` in nextpnr only surfaces results for ice40. 

A workaround I've been using is  `os.environ["AMARANTH_nextpnr_opts"] = "--freq 66.67 -r -v"`

Reply:
@gatecat Do you think that's something we can have shipped in nextpnr soonish? It's fairly hard to convince myself to add support for the `--freq` workaround.

Reply:
The `create_clock` command should work in a PDC file - it seems like the docs are just out of date, I'll fix them tomorrow.

Clocks should also be automatically propagated through PLLs when that is used.

Reply:
Just to confirm, that works with Radiant as well, right?